### PR TITLE
[GOBBLIN-1552] determine flow status correctly when dag manager is disabled

### PR DESCRIPTION
**Pull request by arjun4084346**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1552

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
flow status is determined by looking at the flow level events. but flow level events are not emitted outside of dag manager. so currently flow status is not being determined correctly when dag manager is disabled.
this pr adds a field `flowStatus` to FlowStatus class, which is determined by JobStatus'es when DagManager is not enabled. in a way it used to be done prior to https://github.com/apache/gobblin/commit/59f3beeea1e62dcc1f3901f082e46bc3d237f9fb#diff-141f30f7aa658e9e18316e3e5ecea3b92865dcc88a4c0809b6d5c0ef1598e410 

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
a couple of new unit tests are added. existing test cases are updated

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-service/src/test/java/org/apache/gobblin/service/monitoring/JobStatusRetrieverTest.java [hunk 1]
```diff
@@ -218,6 +222,40 @@ public void testGetLatestExecutionIdsForFlow() throws Exception {
     Assert.assertEquals(this.jobStatusRetriever.getLatestExecutionIdForFlow(FLOW_NAME, FLOW_GROUP), -1L);
   }
 
+  @Test (dependsOnMethods = "testGetLatestExecutionIdsForFlow")
+  public void testGetJobStatusesForFlowExecutionWithDagManager() throws Exception {
```
Comment (phet) [inline gobblin-service/src/test/java/org/apache/gobblin/service/monitoring/JobStatusRetrieverTest.java:226]: do we expect the same result whether or not the DAG Manager is enabled?  if so, should we initialize the status once in common and then have each JobStatusRetriever analyze the same thing (to get the same result)?

File: gobblin-runtime/src/main/java/org/apache/gobblin/service/monitoring/JobStatusRetriever.java [hunk 2]
```diff
@@ -212,4 +224,60 @@ public static boolean isFlowStatus(org.apache.gobblin.service.monitoring.JobStat
     return jobStatus.getJobName() != null && jobStatus.getJobGroup() != null
         && jobStatus.getJobName().equals(JobStatusRetriever.NA_KEY) && jobStatus.getJobGroup().equals(JobStatusRetriever.NA_KEY);
   }
+
+  public ExecutionStatus getFlowStatusFromJobStatuses(Iterator<JobStatus> jobStatusIterator) {
```
Comment (phet) [inline gobblin-runtime/src/main/java/org/apache/gobblin/service/monitoring/JobStatusRetriever.java:228]: may not need to be an instance method.  now it's only for `dagManagerEnabled`, but that state doesn't change, and could anyway be passed as a param (e..g to a static method).  even clearer, perhaps, would cast the flow status analysis/extraction in the Strategy Pattern.

File: gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-server/src/main/java/org/apache/gobblin/service/FlowExecutionResourceLocalHandler.java [hunk 3]
```diff
@@ -202,7 +203,7 @@ public static FlowExecution convertFlowStatus(org.apache.gobblin.service.monitor
         .setExecutionStatistics(new FlowStatistics().setExecutionStartTime(getFlowStartTime(monitoringFlowStatus))
             .setExecutionEndTime(flowEndTime))
```
[the remark this answers is not in the document]
Comment (arjun4084346) [reply gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-server/src/main/java/org/apache/gobblin/service/FlowExecutionResourceLocalHandler.java:204]: Yes that's the assumption. What can make this assumption wrong?

File: gobblin-runtime/src/main/java/org/apache/gobblin/service/monitoring/JobStatusRetriever.java [hunk 4]
```diff
@@ -212,4 +224,60 @@ public static boolean isFlowStatus(org.apache.gobblin.service.monitoring.JobStat
...
+
+    // if any job is cancelled or flow has failed then return failed status
+    if (currentFlowExecutionStatus == ExecutionStatus.CANCELLED ||
+        jobExecutionStatus == ExecutionStatus.CANCELLED) {
+      return ExecutionStatus.CANCELLED;
+    }
+
+    if (currentFlowExecutionStatus == ExecutionStatus.COMPLETE &&
+        jobExecutionStatus == ExecutionStatus.PENDING) {
+      return ExecutionStatus.PENDING;
```
[the remark this answers is not in the document]
Comment (arjun4084346) [reply gobblin-runtime/src/main/java/org/apache/gobblin/service/monitoring/JobStatusRetriever.java:270]: Yea, it was incorrectly copied from https://github.com/apache/gobblin/commit/59f3beeea1e62dcc1f3901f082e46bc3d237f9fb#diff-141f30f7aa658e9e18316e3e5ecea3b92865dcc88a4c0809b6d5c0ef1598e410 where initial status was COMPLETE
I will fix it

File: gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-server/src/main/java/org/apache/gobblin/service/FlowExecutionResourceLocalHandler.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (phet) [reply gobblin-restli/gobblin-flow-config-service/gobblin-flow-config-service-server/src/main/java/org/apache/gobblin/service/FlowExecutionResourceLocalHandler.java:217]: I don't have a specific reason to suspect it's wrong.  if you have an idea on why the assumption is valid, helpful to add a code comment to explain, but if you're also not sure, perhaps wrap the update with `Math.max()`

File: gobblin-service/src/test/java/org/apache/gobblin/service/monitoring/MysqlJobStatusRetrieverTestWithoutDagManager.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    super.testGetJobStatusesForFlowExecution1();
+  }
+
+  @Test (dependsOnMethods = "testGetJobStatusesForFlowExecution1")
+  public void testGetLatestExecutionIdsForFlow() throws Exception {
+    super.testGetLatestExecutionIdsForFlow();
+  }
+
+  @Test (dependsOnMethods = "testGetLatestExecutionIdsForFlow")
+  public void testGetFlowStatusFromJobStatuses() throws Exception {
```
Comment (phet) [inline gobblin-service/src/test/java/org/apache/gobblin/service/monitoring/MysqlJobStatusRetrieverTestWithoutDagManager.java:93]: nice side-by-side illustration from the `MysqlJobStatusRetrieverTest`.

File: gobblin-service/src/test/java/org/apache/gobblin/service/monitoring/MysqlJobStatusRetrieverTestWithoutDagManager.java [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    long flowExecutionId = 1237L;
+
+    addJobStatusToStateStore(flowExecutionId, JobStatusRetriever.NA_KEY, ExecutionStatus.COMPILED.name());
+    Assert.assertEquals(ExecutionStatus.$UNKNOWN,
+        jobStatusRetriever.getFlowStatusFromJobStatuses(jobStatusRetriever.dagManagerEnabled, jobStatusRetriever.getJobStatusesForFlowExecution(FLOW_NAME, FLOW_GROUP, flowExecutionId)));
```
Comment (phet) [inline gobblin-service/src/test/java/org/apache/gobblin/service/monitoring/MysqlJobStatusRetrieverTestWithoutDagManager.java:98]: nit: call statically

**Review by phet — COMMENTED** [review]
I like the approach now in the code, and especially in the tests.

File: gobblin-runtime/src/main/java/org/apache/gobblin/service/monitoring/JobStatusRetriever.java [hunk 7]
```diff
@@ -218,4 +228,42 @@ public static boolean isFlowStatus(org.apache.gobblin.service.monitoring.JobStat
...
+        if (JobStatusRetriever.isFlowStatus(jobStatus)) {
+          flowExecutionStatus = ExecutionStatus.valueOf(jobStatus.getEventName());
+        }
+      }
+    } else {
+      Set<ExecutionStatus> jobStatuses = new HashSet<>();
+      while (jobStatusIterator.hasNext()) {
+        JobStatus jobStatus = jobStatusIterator.next();
+        // because in absence of DagManager we do not get all flow level events, we will ignore the flow level events
+        // we actually get and purely calculate flow status based on flow statuses.
```
[the remark this answers is not in the document]
Comment (arjun4084346) [reply gobblin-runtime/src/main/java/org/apache/gobblin/service/monitoring/JobStatusRetriever.java:248]: in future, i actually plan to remove the 'dagManagerDisabled' blocks :)

File: gobblin-runtime/src/main/java/org/apache/gobblin/service/monitoring/JobStatusRetriever.java [hunk 8]
```diff
@@ -218,4 +228,42 @@ public static boolean isFlowStatus(org.apache.gobblin.service.monitoring.JobStat
...
+      if (jobStatuses.contains(ExecutionStatus.FAILED)) {
+        flowExecutionStatus = ExecutionStatus.FAILED;
+      } else if (jobStatuses.contains(ExecutionStatus.CANCELLED)) {
+        flowExecutionStatus = ExecutionStatus.CANCELLED;
+      } else if (jobStatuses.contains(ExecutionStatus.ORCHESTRATED)) {
+        flowExecutionStatus = ExecutionStatus.ORCHESTRATED;
+      } else if (jobStatuses.contains(ExecutionStatus.RUNNING)) {
+        flowExecutionStatus = ExecutionStatus.RUNNING;
+      } else if (jobStatuses.contains(ExecutionStatus.COMPLETE)) {
+        flowExecutionStatus = ExecutionStatus.COMPLETE;
```
[the remark this answers is not in the document]
Comment (arjun4084346) [reply gobblin-runtime/src/main/java/org/apache/gobblin/service/monitoring/JobStatusRetriever.java:263]: Yes, `KafkaJobStatusMonitor` should update the job status

File: gobblin-runtime/src/test/java/org/apache/gobblin/service/monitoring/FlowStatusGeneratorTest.java [hunk 9]
```diff
@@ -134,8 +143,9 @@ public void testGetFlowStatusesAcrossGroup() {
         Arrays.asList(f0jsmDep2)));
   }
 
-  private FlowStatus createFlowStatus(String flowGroup, String flowName, long flowExecutionId, List<JobStatus> jobStatuses) {
-    return new FlowStatus(flowName, flowGroup, flowExecutionId, jobStatuses.iterator());
+  private FlowStatus createFlowStatus(String flowGroup, String flowName, long flowExecutionId, List<JobStatus> jobStatuses, JobStatusRetriever jobStatusRetriever) {
+    return new FlowStatus(flowName, flowGroup, flowExecutionId, jobStatuses.iterator(),
+        jobStatusRetriever.getFlowStatusFromJobStatuses(jobStatusRetriever.dagManagerEnabled, jobStatuses.iterator()));
```
[the remark this answers is not in the document]
Comment (arjun4084346) [reply gobblin-runtime/src/test/java/org/apache/gobblin/service/monitoring/FlowStatusGeneratorTest.java:148]: final param?

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/FsJobStatusRetriever.java [hunk 10]
```diff
@@ -60,7 +62,8 @@
 
   @Inject
   public FsJobStatusRetriever(Config config, MultiContextIssueRepository issueRepository) {
-    super(issueRepository);
+    super(ConfigUtils.getBoolean(config, ServiceConfigKeys.GOBBLIN_SERVICE_DAG_MANAGER_ENABLED_KEY,
+        ServiceConfigKeys.DEFAULT_GOBBLIN_SERVICE_DAG_MANAGER_ENABLED), issueRepository);
```
[the remark this answers is not in the document]
Comment (arjun4084346) [reply gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/FsJobStatusRetriever.java:66]: Yes, I would prefer to leave it like this. If in future we need more params to be passed, we again have to change the signature/DI logic. We already need other configs in the next line to create a state store

File: gobblin-service/src/test/java/org/apache/gobblin/service/monitoring/JobStatusRetrieverTest.java [hunk 11]
```diff
@@ -103,8 +103,8 @@ public void testGetJobStatusesForFlowExecution() throws IOException {
     long flowExecutionId = 1234L;
     addJobStatusToStateStore(flowExecutionId, JobStatusRetriever.NA_KEY, ExecutionStatus.COMPILED.name());
 
-    Iterator<JobStatus>
-        jobStatusIterator = this.jobStatusRetriever.getJobStatusesForFlowExecution(FLOW_NAME, FLOW_GROUP, flowExecutionId);
+    List<JobStatus> jobStatuses = ImmutableList.copyOf(this.jobStatusRetriever.getJobStatusesForFlowExecution(FLOW_NAME, FLOW_GROUP, flowExecutionId));
```
[the remark this answers is not in the document]
Comment (arjun4084346) [reply gobblin-service/src/test/java/org/apache/gobblin/service/monitoring/JobStatusRetrieverTest.java:106]: getJobStatusesForFlowExecution is still non-static

File: gobblin-service/src/test/java/org/apache/gobblin/service/monitoring/MysqlJobStatusRetrieverTestWithoutDagManager.java (hunk 6 above)
Comment (arjun4084346) [reply gobblin-service/src/test/java/org/apache/gobblin/service/monitoring/MysqlJobStatusRetrieverTestWithoutDagManager.java:98]: maybe it was static in b/w changes. not now

File: gobblin-runtime/src/test/java/org/apache/gobblin/service/monitoring/FlowStatusGeneratorTest.java [hunk 12]
```diff
@@ -106,10 +108,13 @@ public void testGetFlowStatusesAcrossGroup() {
         f0Js2Status, f0Js2Tag, f0Js2JobGroup1, f0Js2JobName1, JOB_EXEC_ID);
 
     // IMPORTANT: result invariants to honor - ordered by ascending flowName, all of same flowName adjacent, therein descending flowExecutionId
-    // NOTE: `Supplier`/thunk needed for repeated use, due to mutable, non-rewinding `Iterator FlowStatus.getJobStatusIterator`
-    Supplier<FlowStatus> createFs1 = () -> createFlowStatus(flowGroup, flowName1, flowExecutionId1, Arrays.asList(f1Js0, f1Js1, f1Js2));
-    Mockito.when(jobStatusRetriever.getFlowStatusesForFlowGroupExecutions(flowGroup, countPerFlowName)).thenReturn(
-        Arrays.asList(createFs1.get()), Arrays.asList(createFs1.get()), Arrays.asList(createFs1.get())); // (for three invocations)
+    // NOTE: Three copies of FlowStatus are needed for repeated use, due to mutable, non-rewinding `Iterator FlowStatus.getJobStatusIterator`
+//    Mockito.when(JobStatusRetriever.getFlowStatusFromJobStatuses(anyBoolean(), any())).thenReturn(ExecutionStatus.RUNNING);
```
Comment (phet) [inline gobblin-runtime/src/test/java/org/apache/gobblin/service/monitoring/FlowStatusGeneratorTest.java:112]: did this commented-out code sneak in?

**Review by phet — APPROVED** [review]
looks good!

**Review by ZihanLi58 — APPROVED** [review]
+1

**Pull request merged**